Only in mbedtls-2.7.5/include/mbedtls: #bn_mul.h#
diff -ur mbedtls-2.7.5/include/mbedtls/bn_mul.h mbedtls-2.7.5.patch/include/mbedtls/bn_mul.h
--- mbedtls-2.7.5/include/mbedtls/bn_mul.h	2018-07-26 15:33:14.000000000 +0200
+++ mbedtls-2.7.5.patch/include/mbedtls/bn_mul.h	2018-08-22 14:37:31.000000000 +0200
@@ -56,7 +56,7 @@
  * This is done as the number of registers used in the assembly code doesn't
  * work with the -O0 option.
  */
-#if defined(__i386__) && defined(__OPTIMIZE__)
+#if defined(__i386__) && defined(__OPTIMIZE__) && (!defined(__ANDROID_API__) || defined(__clang__))
 
 #define MULADDC_INIT                        \
     asm(                                    \
